mod-login-saml/src/main/resources/domain/mod-login-saml/features/loginSamlUnauthenticated.feature

```diff
@@ -0,0 +1,17 @@
+# Tests of unauthenticated routes (no okapi token) don't require the background variables so we add it to a different
+# feature here.
+Feature: Login SAML Unauthenticated
+
+  Background:
+    * url baseUrl
+    * configure headers = { 'x-okapi-tenant': #(testTenant) }
+
+  @Undefined
+  Scenario: Check endpoint bad request
+  * print 'undefined'
+# TODO In the rest assured tests, this is 400, not 200 active false. I believe this is because when the tests
+# are run against vertx directly the request without the headers responds 400 (bad request) -- never reaching
+# getSamlCheck. But when it is run through okapi the request is routed to the getSamlCheck method.
```

Comment:
Yes and no. RMB generated API implementation checks for `x-okapi-tenant` header. If updating the rest assured test with `x-okapi-tenant` header, it will return 200 as well.

Reply:
Actually the ```x-okapi-tenant``` is there in the new karate test (specified in the ````background``` part). This is what the request response looks like:

```
18:28:00.641 request:
1 > GET https://folio-testing-okapi.dev.folio.org:443/saml/check
1 > Accept-Encoding: gzip,deflate
1 > Connection: Keep-Alive
1 > Host: folio-testing-okapi.dev.folio.org:443
1 > User-Agent: Apache-HttpClient/4.5.12 (Java/11.0.10)
1 > x-okapi-tenant: testenant48810419905537024

18:28:00.785 response time in milliseconds: 143.34
1 < 200
1 < Connection: keep-alive
1 < Content-Type: application/json
1 < Date: Fri, 11 Jun 2021 22:28:01 GMT
1 < Transfer-Encoding: chunked
1 < X-Okapi-Trace: [GET mod-authtoken-2.9.0-SNAPSHOT.91 http://10.36.1.135:9134/saml/check : 202 544us, GET mod-login-saml-2.2.1-SNAPSHOT.70 http://10.36.1.135:9146/saml/check : 200 15718us]
1 < vary: origin
{
  "active": false
}
```
I think what is happening is that the old rest assured test doesn't include the  ```x-okapi-header``` and returns 400 bad request because it's missing the header (there's no method/route for it to hit) (that's [what the test wants](https://github.com/folio-org/mod-login-saml/blob/d8e202e0a307d12b2080dd5b76a7afd59463c8d9/src/test/java/org/folio/rest/impl/SamlAPITest.java#L136)). But since we're now going through okapi, we need that header otherwise we'll get 404. Not sure if this test for 400 is really that useful since I think it's only testing the vertx endpoint. I'd argue it could be removed.

Reply:
Agreed. The old rest assured tests are not integration tests. It is not necessary/valid to translate all old rest assured tests to Karate integration tests.

Reply:
This has got me thinking that maybe we should hold off merging this until we've had a chance to discuss more what we're really trying to accomplish with the karate tests. There might be a bigger issue here than this particular test.

My assumption has been that we're wanting to fully replace the rest assured tests with karate tests. So my "test plan" (the ```@Undefined``` tests in this commit) are mostly a 1 for 1 match between the tests in SamlAPITests.java, without really fully understanding why. So someone coming to work on the next phase (implementing the tests) would also assume that is the goal -- which could lead to a lot of unnecessary work.

Looking at the other two modules which have created their test plans (the other [closed tickets here](https://wiki.folio.org/pages/viewpage.action?spaceKey=FOLIJET&title=Karate+API+Integration+Testing)), it would appear that they are operating under the same assumption -- that it should be a 1 for 1 replacement of rest assured with karate. See [this](https://github.com/folio-org/folio-integration-tests/blob/master/mod-patron-blocks/src/main/resources/domain/mod-patron-blocks/features/automatedPatronBlocks.feature) and [this](https://github.com/folio-org/mod-patron-blocks/blob/master/src/test/java/org/folio/rest/impl/AutomatedPatronBlocksAPITest.java) for what I mean. We're just replicating the tests between the two systems, which may be of questionable benefit.

I think for example instead of completely replacing the rest assured tests with karate, an argument could be made to just consider karate "another tool in our toolbelt" -- making it easier to do cross-module integration tests when they are needed, or UI tests, or whatever karate does well. If that's the case, these test plans that we're making would look  very different, and also the amount of work required to implement karate tests would be quite different too.

Maybe this discussion has already taken place and me and the other folks working on the test plans just aren't aware of the background, but also maybe we're rushing into something that isn't that great of an idea.

Reply:
Found this [wiki page](https://wiki.folio.org/display/DD/Karate+API+Integration+Tests+Implementation+Guidelines) that might be helpful. It mentions the scope of Karate integration testing
```
Scope 

1. - Identify most important business flows
2. - Add test case to cover a bug condition
3. - Don't write tests for all possible conditions 

```

Reply:
Thanks Hongwei for the background material. It does seem like maybe there's some recognition in this that replicating all the http tests would be overkill. But then again it seems like people are going down that road. We should discuss more maybe in our call tomorrow.

If we agree that doing a 1 for 1 replacement is a bad idea there might be an opportunity to clarify what would constitute a "good use of a karate integration test". Maybe we just need a test or two that captures the whole end to end chain (tenant creation/user create/permissions/login etc) -- which is what these tests are all about.

Reply:
The goal of testing is to find as many bugs as possible.
Converting existing tests into Karate tests is unlikely to find more bugs.
During bugfest many librarians manually test workflows - these tests might be automated.

There are different levels of testing: https://en.wikipedia.org/wiki/Software_testing#Testing_levels
We have some integration tests at the module level and maven has an integration-test phase for this purpose. These integration tests should stay at the module level for earliest feedback.
Karate is for system testing, and for integration testing in areas no covered by module based integration tests.

Reply:
Thanks Julien, Hongwei and Adam for your thoughts on this and for the time this morning to discuss. I think we're agreed that we don't want to replace all the rest assured tests with karate but rather want the benefit of both frameworks. I revised the test plan to cover what I think is the critical part of saml to test (a handful of login routes). Let me know if you feel there are others that could benefit from karate but this might be a good start.